New notebook launch ux

zarsky-broad opened the pull request:
resolves #1089 
Note: the weird state here is if you specifically open a notebook read-only in a workspace that you have compute and write permissions in, but with no cluster. Then, you can click 'edit in Jupyter' and have to click 'create cluster' afterwards. This requires a larger refactor to address.

zarsky-broad committed:
the no cluster case

zarsky-broad committed:
handle creation/resuming

zarsky-broad committed:
cleanup & simplify

zarsky-broad committed:
dedupe cluster list filter

zarsky-broad committed:
Merge branch 'dev' into iz-new-notebook-launch-ux

zarsky-broad committed:
don't need this once...

zarsky-broad committed:
working on terminal launcher...

zarsky-broad committed:
Merge branch 'dev' into iz-new-notebook-launch-ux

zarsky-broad committed:
track creation request for terminal

zarsky-broad committed:
accidentally a const

elizzhao commented on src/pages/workspaces/workspace/notebooks/TerminalLauncher.js:
```diff
@@ -46,33 +45,69 @@ const TerminalLauncher = _.flow(
+      return div({ style: { padding: '2rem' } }, [
+        spinner({ style: { color: colors.green[0], marginRight: '0.5rem' } }),
+        (clusterStatus === 'Creating' || !cluster || createRequested) ?
+          'Creating notebook runtime environment. You can navigate away and return in 5-10 minutes.' :
```
Should these still say "notebook" for the terminal?

katevoss replied:
```suggestion
          'Creating runtime environment. You can return in 5-10 minutes.' :
```
Same clarification as above

katevoss commented on src/pages/workspaces/workspace/notebooks/NotebookLauncher.js:
```diff
@@ -347,42 +322,79 @@ class NotebookEditor extends Component {
+          ]) :
+          div({ style: { padding: '2rem', fontSize: 16, fontWeight: 'bold' } }, [
+            'You are viewing this notebook in read-only mode. You can ',
+            linkButton({ onClick: () => this.setState({ createOpen: true }) }, 'create a notebooks runtime'),
```
On this branch:
<img width="1659" alt="Screen Shot 2019-04-03 at 10 16 32 AM" src="https://user-images.githubusercontent.com/2395211/55486027-97973e80-55f9-11e9-9627-d89fe7628772.png">
There is no `x` to leave the notebook. I think this should be re-added so a user can navigate back to their list of Notebooks.
Also I think we should keep the current copy, "Viewing read-only" and make the `Create runtime` as a button outline, less text is clearer.

katevoss commented on src/pages/workspaces/workspace/notebooks/NotebookLauncher.js:
```diff
@@ -347,42 +322,79 @@ class NotebookEditor extends Component {
+                  'Creating notebook runtime environment. You can navigate away and return in 5-10 minutes.' :
+                  'Starting notebook runtime environment, this may take up to 2 minutes.'
+                ),
+                makeStep(2, 'Copy notebook', localizeFailures ?
```
This looks like a button, now that we have the `buttonOutline` style. I tried clicking on it for 2 minutes. Can you make this more clearly a step in the process?

elizzhao replied:
"Copy notebook" seems like you could click it and copy the notebook to paste somewhere else. Perhaps a more descriptive wording would clear up that confusion?

katevoss commented on src/pages/workspaces/workspace/notebooks/NotebookLauncher.js:
```diff
@@ -347,42 +322,79 @@ class NotebookEditor extends Component {
+              ]) :
+              h(Fragment, [
+                makeStep(1, 'Runtime started', (isCreating || createRequested) ?
+                  'Creating notebook runtime environment. You can navigate away and return in 5-10 minutes.' :
```
```suggestion
                  'Creating notebook runtime environment. You can return in 5-10 minutes.' :
```
Saves button real-estate

elizzhao commented on src/pages/workspaces/workspace/notebooks/TerminalLauncher.js:
```diff
@@ -46,33 +45,69 @@ const TerminalLauncher = _.flow(
+      return div({ style: { padding: '2rem' } }, [
+        spinner({ style: { color: colors.green[0], marginRight: '0.5rem' } }),
+        (clusterStatus === 'Creating' || !cluster || createRequested) ?
+          'Creating notebook runtime environment. You can navigate away and return in 5-10 minutes.' :
```
![Screen Shot 2019-04-03 at 11 15 25 AM](https://user-images.githubusercontent.com/42386483/55490516-e3e67c80-5601-11e9-9c33-58b5ae8e4991.png)

The message doesn't update even when the cluster is running. I had this window open for a while, so I thought it was odd that it still said "creating runtime environment" (also since the cluster said it was running). When I refreshed the page, the terminal opened immediately. I think it may be an updating issue?

elizzhao commented on src/pages/workspaces/workspace/notebooks/NotebookLauncher.js:
```diff
@@ -347,42 +322,79 @@ class NotebookEditor extends Component {
+      const isRunning = clusterStatus === 'Running'
+      const currentStep = isRunning ? 2 : 1
+
+      const makeStep = (index, shortText, fullText) => {
```
@katevoss and I were chatting about ideas to maybe make it clearer that these are steps. I think it could be useful to add 2 dots in between each step. It adds clarity that these are steps that happen one after the other, and it also mimics the "Script/Inputs/Outputs/Run Analysis" flow in the workflow view. 

thoughts?

panentheos replied:
Since there are only two steps, and the second is typically pretty fast, I wonder how much value users actually get out of seeing them laid out. Perhaps we could just show a single message with A) what's currently happening, and B) how long we estimate until the notebook is ready.

panentheos commented:
Currently if you click the cluster manager to create a cluster, instead of the link, it never localizes and spins indefinitely on the 'copying' step.

elizzhao committed:
include public workspaces

elizzhao committed:
Revert "include public workspaces"

This reverts commit 1cd9c8d44c70875a3a8429328d03d21eda61eeb4.

elizzhao commented:
(sorry i messed up and accidentally committed to your branch instead of mine oops)

zarsky-broad committed:
Merge branch 'dev' into iz-new-notebook-launch-ux

katevoss reviewed (commented):
See comments

zarsky-broad committed:
working on cleanup

zarsky-broad committed:
Merge branch 'dev' into iz-new-notebook-launch-ux

zarsky-broad committed:
Merge branch 'dev' into iz-new-notebook-launch-ux

zarsky-broad committed:
cleanup & pr comments

zarsky-broad committed:
simplify...

zarsky-broad committed:
unnecessary helper

zarsky-broad committed:
fixed non-nav cluster refresh

zarsky-broad committed:
terminal working now?

panentheos commented on src/components/ClusterManager.js:
```diff
@@ -517,16 +512,19 @@ export default ajaxCaller(class ClusterManager extends PureComponent {
-      }, [icon('terminal', { className: 'is-solid', size: 24 })]),
+        )
+      }, [
+        linkButton({
```
This should be a `link` since it uses `href`. `linkButton` is intended for buttons (e.g. using `onClick`) that have the visual appearance of links.

panentheos commented on src/pages/workspaces/workspace/notebooks/NotebookLauncher.js:
```diff
@@ -144,30 +89,45 @@ class NotebookPreviewFrame extends Component {
+  url: undefined,
+  saving: false,
+  createOpen: false,
+  createRequested: false, // to handle the pause while Leo starts creation
```
Is this in response to the fact that newly created clusters don't show up in subsequent 'list' requests for a few seconds? If so, that's a known Leo bug that should be fixed at some point. If we're going to compensate for it in the short term, I think we should put a larger comment here spelling out why we're doing this and when it can be removed.

zarsky-broad replied:
Apparently this is fixed now!

panentheos commented on src/pages/workspaces/workspace/notebooks/NotebookLauncher.js:
```diff
@@ -179,43 +139,78 @@ class NotebookEditor extends Component {
-          break
-        default:
-          console.log('Unrecognized message:', e.data)
+      if (!e.data.source.startsWith('react-')) { // for the react dev tools
```
This crashes because not all messages follow this shape.

Rather than special-casing this, we should remove the `console.log` below, since all kinds of messages can be passed through this channel.

panentheos commented on src/pages/workspaces/workspace/notebooks/NotebookLauncher.js:
```diff
@@ -252,50 +248,24 @@ class NotebookEditor extends Component {
         refreshClusters()
         await Utils.delay(10000)
+      } else if (status === 'Creating') {
+        await Utils.delay(15000)
```
TOL: These intervals seem arbitrary and somewhat fiddly. Can we come up with a more unified approach here?

zarsky-broad committed:
Merge branch 'dev' into iz-new-notebook-launch-ux

zarsky-broad committed:
trying merge again...

panentheos commented on src/pages/workspaces/workspace/notebooks/TerminalLauncher.js:
```diff
@@ -18,61 +20,86 @@ const TerminalLauncher = _.flow(
+        await Utils.delay(15000)
+      } else {
+        await Utils.delay(3000)
+      }
```
Can we factor this out since it's the same logic as the other launcher?

panentheos commented on src/pages/workspaces/workspace/notebooks/TerminalLauncher.js:
```diff
@@ -18,61 +20,86 @@ const TerminalLauncher = _.flow(
+    }
 
-    this.scheduledRefresh = setTimeout(() => this.refreshCookie(clusterName), 1000 * 60 * 20)
+    while (this.mounted) {
```
I don't think it's necessary for us to maintain `this.mounted` manually since we're cancelling ajax requests implicitly.

zarsky-broad replied:
But wouldn't just go forever if I just did `while (true)`? (at least, forever until the cluster returned 'running')

panentheos replied:
Nope, if the ajax call gets cancelled the whole process should stop running.

panentheos commented on src/pages/workspaces/workspace/notebooks/TerminalLauncher.js:
```diff
@@ -18,61 +20,86 @@ const TerminalLauncher = _.flow(
+    await refreshClusters()
+    if (!this.props.cluster) {
+      await Jupyter.cluster(namespace, Utils.generateClusterName()).create({
+        machineConfig: Utils.normalizeMachineConfig({})
```
What about providing the same cluster create affordance here as on the other launcher?

zarsky-broad replied:
Brad's good with just simple autocreate here.

zarsky-broad committed:
pr comments 1

zarsky-broad committed:
pr comments 2

zarsky-broad committed:
pr comments 3

Note: the weird state here is if you specifically open a notebook read-only in a workspace that you have compute and write permissions in, but with no cluster. Then, you can click 'edit in Jupyter' and have to click 'create cluster' afterwards. This requires a larger refactor to address.

panentheos commented on src/libs/utils.js:
```diff
@@ -281,3 +281,20 @@ export const withBusyState = _.curry((setBusy, fn) => async (...args) => {
+    case 'Creating':
+      return delay(15000)
+    default:
+      return delay(3000)
```
FYI, while testing this I noticed that we have a parallel refresh process in `ClusterManager`, so we're actually making twice as many calls in total. Not something we should fix here, but good to keep in mind for a larger round-up of this code later.

zarsky-broad replied:
Legit.

panentheos reviewed (approved):
Looks good. There's still more we could do for this page, but this is a solid improvement.

zarsky-broad committed:
Merge branch 'dev' into iz-new-notebook-launch-ux
